### ISSUE TITLE
Think about the API

### ISSUE THREAD

**Issue by paultag** [paultag/go-resolver#3]
Figure out what the API should look like here

**Comment by tianon** [comment]
+1, it feels like it was thrown together pretty ad-hoc right now (although it's totally workable)

**Comment by tianon** [comment]
IMO we should also think about whether this module itself should handle things like topsort and pinning resolution, or whether it needs to export something that can be used for those types of things without too much trouble.

**Comment by paultag** [comment]
I'd say pinning (perhaps not by that name, since it already has meaning)
yes, topsort maybe later :)

**Comment by paultag** [comment]
Which means candidate needs to grow a lot (or not?)

**Comment by tianon** [comment]
Hmm, not sure.  We definitely need some notion of "priority" in each `BinaryIndex`, and then the candidates list either needs to know that priority (maybe just a full ref to the `BinaryIndex` the package came from?) or we need the candidates list to be inherently sorted by priority, which is harder, I think.  It would also be a lot more work, because we'd be sorting the full list based on priority when it only matters after we've got a final set of candidates that satisfy a condition.

**Comment by paultag** [comment]
Yeah. Let's add priority complexity here - perhaps wrap the binary index?
Keep them in priority order on another Indexes super object w/ helpers?

**Comment by tianon** [comment]
That seems reasonably sane to me! :+1:
